openjdk/jfx
# 8328629: JUnit test without a timeout value can hang indefinitely

This PR adds a default timeout of 120 seconds for JUnit 5 tests that do not have an explicit `@Timeout` on either the methods or the class, and a default timeout of 20 seconds for lifecycle methods (e.g., `@BeforeEach`, `@BeforeAll`, `@AfterEach`, `@AfterAll`).

JUnit 5 effects its timeout by interrupting the test thread. We have several tests that catch and ignore InterruptedException for sleep or await, so I fixed these tests to throw an error if InterruptedException is caught (they may or may not be a problem, but if any of them can happen in a loop, it would block the timeout from killing the test).

While testing this, I discovered that by itself, this JUnit 5 timeout is ineffective if the deadlock described in [JDK-8238505](https://bugs.openjdk.org/browse/JDK-8238505)  occurs. The reason for this is that the deadlock almost always occurs when called from a shutdown hook (called by `System.exit`), and that prevents the process from exiting even when interrupted by the test runner after the timeout. A separate PR is out for that bug fix.

### Testing instructions

I tested this in connection with PR #1574 / [JDK-8340405](https://bugs.openjdk.org/browse/JDK-8340405) using my [test-timeout-and-shutdown-hang](https://github.com/kevinrushforth/jfx/tree/test-timeout-and-shutdown-hang) branch. That branch has the fixes from both this PR and PR #1574 as well as tests that will hang indefinitely without the fixes and fail due to the expected timeout with the fixes.

The default timeout values for test and lifecycle methods can be modified using a gradle property:

* `JUNIT_TEST_TIMEOUT` : default = 120s
* `JUNIT_LIFECYCLE_TIMEOUT` : default = 20s

For example, to set a timeout of 60 seconds for tests and 15 seconds for lifecycle methods:

```
gradle -PJUNIT_TEST_TIMEOUT=60s -PJUNIT_LIFECYCLE_TIMEOUT=15s test
```


---------
### Progress

### Issue
 * [JDK-8328629](https://bugs.openjdk.org/browse/JDK-8328629): JUnit test without a timeout value can hang indefinitely (**Bug** - P4)


### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**) Review applies to [e16ef7ec](https://git.openjdk.org/jfx/pull/1575/files/e16ef7ece7cb95ab0baf2fd37349396e4986fcbf)
 * [Marius Hanl](https://openjdk.org/census#mhanl) (@Maran23 - Committer) Review applies to [8a4078fe](https://git.openjdk.org/jfx/pull/1575/files/8a4078fed1941bdd7052c1271079c4bfed378fba)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1575/head:pull/1575` \
`$ git checkout pull/1575`

Update a local copy of the PR: \
`$ git checkout pull/1575` \
`$ git pull https://git.openjdk.org/jfx.git pull/1575/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1575`

View PR using the GUI difftool: \
`$ git pr show -t 1575`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1575.diff">https://git.openjdk.org/jfx/pull/1575.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1575#issuecomment-2361894421)
